### PR TITLE
Generate cases for all ScalarTypes in Type functions that call to TH

### PR DESCRIPTION
**Pull request by li-roy**
This revamps the way functions that call into TH are generated. Now, CPUFloatType, CPUIntType, etc. all generate these functions identically. The next step is to deduplicate these into a single class.

note: This PR introduces a lot of redundant codegen, so I won't merge it until subsequent PRs that deduplicate the Type classes are ready.

### CODE REVIEW

File: aten/src/ATen/function_wrapper.py [hunk 1]
```diff
@@ -608,103 +632,125 @@ def to_return_type(arg, option):
...
-        if default is None:
-            # cause the default constructor for the object to run
-            return '{}'
-        if 'if_true' in argument:
-            return argument['default'] == argument['if_true']
-        for pattern, replacement in HEADER_CONSTANT_REPLACEMENTS:
-            default = re.sub(pattern, replacement, str(default))
-        if type_str in {'Scalar', 'int64_t', 'double'}:
+# translates defaults from cwrap types to C++ values
+def translate_default(argument, type_str, default):
```
Comment (li-roy) [inline aten/src/ATen/function_wrapper.py]: It looks like there's a lot of changes here, but it's really just moving `translate_default`, `translate_formal`, `get_formals`, `find_dispatch_tensor` out of `create_generic` so I can use them elsewhere.

Comment (ezyang) [reply aten/src/ATen/function_wrapper.py]: It'll be a lot easier to review the PR if you give us a billing of changes. Just a blow-by-blow, "Here are the logical units of change I made in this PR". Then we'll know what to expect in review.

**Review by gchanan — COMMENTED** [review]
can we call our existing dispatch macros instead of generating new ones?

I'm presuming there are some combinations of scalar types that we don't have dispatch macros for, but I'd want to know what those are.

**Review by ezyang — CHANGES_REQUESTED** [review]
dismiss me when billing of changes is added

**Pull request closed without merging**